### PR TITLE
python-tk: Separate tkinter from python@3.9

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This trims the large `tcl-tk` dep tree on Linux, which pulls in a number of X11 dependencies. It will also resolve the cyclic dependency issue in `xcb-proto`.

### CODE REVIEW

**Review by MikeMcQuaid — DISMISSED** [review]

File: Formula/python@3.9.rb [hunk 1]
```diff
@@ -457,6 +457,9 @@ def caveats
       They will install into the site-package directory
         #{HOMEBREW_PREFIX/"lib/python#{version.major_minor}/site-packages"}
 
+      tkinter is no longer a part of this formula, but is available separately:
+        brew install python-tk@3.9
```
Comment (carlocab) [inline Formula/python@3.9.rb:461]:
```suggestion
        brew install python-tk@#{version.major_minor}
```

File: Formula/python@3.9.rb [hunk 2]
```diff
@@ -168,8 +168,8 @@ def install
     # Avoid linking to libgcc https://mail.python.org/pipermail/python-dev/2012-February/116205.html
     args << "MACOSX_DEPLOYMENT_TARGET=#{MacOS.version}"
 
-    args << "--with-tcltk-includes=-I#{Formula["tcl-tk"].opt_include}"
-    args << "--with-tcltk-libs=-L#{Formula["tcl-tk"].opt_lib} -ltcl8.6 -ltk8.6"
+    # Disable _tkinter - this is built in a separate formula
```
Comment (sjackman) [inline Formula/python@3.9.rb:171]:
```suggestion
    # Disable _tkinter - this is built in a separate formula python-tk
```

**Comment by Bo98** [comment]
Since I don't want to make two revision bumps in quick succession, I've added another commit of some more changes I wanted to make. If there are objections, I can move it to a separate PR.

The change involves moving most of the build of `pip`, `setuptools` and `wheel` from postinstall to install. Instead of placing the source contents into `libexec` and running the full install at `postinstall` time, I've changed it so that we build wheels of these and then install that at postinstall time. This leads to a significantly faster `postinstall` (at least 3x faster). It also means we've not got unnecessary test files in our bottles (some of which we already manually removed due to various issues), and will help with an upcoming audit: #10851.

**Comment by sjackman** [comment]
> Instead of placing the source contents into libexec and running the full install at postinstall time, I've changed it so that we build wheels of these and then install that at postinstall time.

Why is it necessary to build the wheel and then install that wheel in postinstall, rather than just install it in the first place?

> This leads to a significantly faster postinstall (at least 3x faster).

Yeah!

**Comment by Bo98** [comment]
> Why is it necessary to build the wheel and then install that wheel in postinstall, rather than just install it in the first place?

Because these are installed outside of the Cellar.

**Comment by danielnachun** [comment]
> The change involves moving most of the build of `pip`, `setuptools` and `wheel` from postinstall to install. Instead of placing the source contents into `libexec` and running the full install at `postinstall` time, I've changed it so that we build wheels of these and then install that at postinstall time. This leads to a significantly faster `postinstall` (at least 3x faster). It also means we've not got unnecessary test files in our bottles (some of which we already manually removed due to various issues), and will help with an upcoming audit: [Homebrew/brew#10851](https://github.com/Homebrew/brew/pull/10851).

Would this strategy also work for other Python packages?  If we can just build wheels during install and then install them to the Cellar during postinstall that would reduce the install time for other packages with Python resources as well.

**Comment by carlocab** [comment]
Not quite. Other Python-based formulae are installed from bottles, so all that's really happening is an untar operation (and maybe some relocation). pip, setuptools, and wheel are different, because they are currently built in the post-install step, which takes a bit longer than an untar + relocation.

**Comment by danielnachun** [comment]
> Not quite. Other Python-based formulae are installed from bottles, so all that's really happening is an untar operation (and maybe some relocation). pip, setuptools, and wheel are different, because they are currently built in the post-install step, which takes a bit longer than an untar + relocation.

Ah right, I guess this only comes up if there is a post-install step which is building wheels but I guess we don't do that for other formula.

**Comment by carlocab** [comment]
Failures are the usual suspects, except perhaps on Big Sur, but those appear to all be out-of-space errors.

Big Sur (ARM)

```
Error: 1 failed step!
brew test --retry --verbose cargo-edit
```

Big Sur (Intel)
```
Error: 21 failed steps!
brew test --retry --verbose asuka
brew test --retry --verbose cargo-edit
brew install --only-dependencies creduce
brew install creduce
brew install --only-dependencies crystal
brew install crystal
brew install --only-dependencies crystal-icr
brew install crystal-icr
brew test --retry --verbose hyperkit
brew install llvm@7
brew install llvm@8
brew install llvm@9
brew install --only-dependencies oclgrind
brew install oclgrind
brew install root
brew install swift
brew install tesseract-lang
brew install --only-dependencies --include-test tree-sitter
brew test --retry --verbose tree-sitter
brew test --retry --verbose wasm-pack
brew install widelands
```

Catalina
```
Error: 3 failed steps!
brew test --retry --verbose cargo-edit
brew test --retry --verbose hyperkit
brew test --retry --verbose wasm-pack
```

Mojave
```
Error: 5 failed steps!
brew test --retry --verbose cargo-edit
brew test --retry --verbose hyperkit
brew test --retry --verbose opencascade
brew test --retry --verbose swift
brew test --retry --verbose wasm-pack
```

**Comment by Bo98** [comment]
> those appear to all be out-of-space errors.

I promise I'll look into this next week. I do have some local changes to `brew` that could fix this.

**Comment by carlocab** [comment]
Looks like most (all?) of my installed formulae that depend on Python still have `tcl-tk` as a runtime dependency in their `INSTALL_RECEIPT.json`. I guess that'll be fixed eventually.

**Comment by Bo98** [comment]
That's an artefact of how the install receipt works. It's useful for formulae that are implicitly linking to `tcl-tk`, but little annoying for those that don't. We saw this in the `openssl@1.1` migration: several formulae were left with a runtime dependency to OpenSSL 1.0 in their install receipt after OpenSSL 1.0 was deleted (and thus prevented users from uninstalling OpenSSL 1.0) - none of which we were aware of as we don't test for that.

**Comment by carlocab** [comment]
Is `INSTALL_RECEIPT.json` checked for dependencies when I install a bottle?

e.g. suppose I install a Python-dependent formula now, which doesn't depend on `tcl-tk`, but its install receipt specifies that dependency. Will `brew` also install `tcl-tk` if it hasn't yet?

**Comment by Bo98** [comment]
No, it won't. In fact, the `runtime_dependencies` is reevaluated on install - so what's stored in the bottle doesn't really matter: https://github.com/Homebrew/brew/blob/679dbea79fcb3b96aa2e21e6665e3cce07f946b3/Library/Homebrew/formula_installer.rb#L804-L809

**Pull request closed without merging**